### DISCUSSION TITLE
Trying to get the realm local address, and verifying the multicast functionality of OTBR

### DISCUSSION THREAD

**Discussion by ZhouJooJoo** [openthread/openthread#9912] [Q&A]
hi there,

I am currently trying to use the samples of "ncs/nrf/samples/openthread/coap_server" and "coap_client". I have flashed them onto two nrf52840dk_nrf52840 development boards. I have called the API to attempt to obtain the realm local multicast address and tried to use OTBR running on a Raspberry Pi to send multicast packets to the realm local multicast address. My ultimate goal is to verify whether messages sent to the realm local address in the Thread network will be resent twice due to the MPL mechanism. At the moment, I have the following questions:

1. When I obtain the realm local multicast address, does it mean that devices in the Thread network composed of coap_server and coap_client have all subscribed to this realm local address? If I send multicast packet information to this address, can they receive it? if they still not subscribe the address, what should i do to let them subscribe it?

2. How can I verify the information sent to the multicast address? The methods I am currently considering are using a dongle to capture packets in the air(sniffer) for verification or printing the received multicast information in the RTT viewer window. However, I lack experience in this and would like to know if there are similar samples that I can refer to.

3. i tried to use the RTT viewer to print the realm local address i got, but the realm local address stays the same for ALL THE TIME, every time they got the same result.
the realm local address will be : 33ff:4000:d9fd:08a6:197b:899a:0000:0100, but as far as i know,  the realm local addrees depneds on the mesh local prefix,Could it be that every time I start two development boards, their mesh local prefixes are the same?Maybe there are some deviations in my understanding. I hope you can correct me. Thank you very much.
i would be appreciate if anyone could answer the questions, many thanks!

best regards,

Zhou

**Comment by jwhui** [comment]
> 1. When I obtain the realm local multicast address, does it mean that devices in the Thread network composed of coap_server and coap_client have all subscribed to this realm local address? If I send multicast packet information to this address, can they receive it? if they still not subscribe the address, what should i do to let them subscribe it?

All devices will automatically subscribe to the realm-local All Thread Nodes address.
Routers will automatically subscribe to the realm-local All Thread Routers address.
All devices must explicitly subscribe to any other realm-local multicast address.

> 2. How can I verify the information sent to the multicast address? The methods I am currently considering are using a dongle to capture packets in the air(sniffer) for verification or printing the received multicast information in the RTT viewer window. However, I lack experience in this and would like to know if there are similar samples that I can refer to.

Packet sniffer would provide visibility into packets transmitted and the IPv6 destination address contained within.

Logs would provide visibility into packets received.

> 3. i tried to use the RTT viewer to print the realm local address i got, but the realm local address stays the same for ALL THE TIME, every time they got the same result.
the realm local address will be : 33ff:4000:d9fd:08a6:197b:899a:0000:0100, but as far as i know, the realm local addrees depneds on the mesh local prefix,Could it be that every time I start two development boards, their mesh local prefixes are the same?Maybe there are some deviations in my understanding. I hope you can correct me. Thank you very much.

Yes, realm-local multicast address is derived from mesh-local prefix.

**Reply by ZhouJooJoo** [reply]
thanks for your reply!  when you mentioned in the second point two ways to verify the multicast message, when you said the packet sniffer, can i use the nrf52840_dongle packet sniffer to capture the data? or if i would like to print the message to the log, should i code some call back function to do that? could you give me some advice or examples about how to deal with the multicast packets? cause i havent't seen any in the samples. thankyou
best wishes

**Reply by jwhui** [reply]
nRF USB Dongle works as a Thread packet sniffer.

OpenThread already logs packets sent/received if you set the log level to INFO or higher.

**Reply by ZhouJooJoo** [reply]
thats awesome, thanks for reply! i will try this method.

**Reply by ZhouJooJoo** [reply]
Hello,

I followed your advice and conducted a series of operations. Currently, the topology of my Thread network is roughly as follows. By using the operating system of OTBR, I can ping the realm local address and receive replies containing information from the local address. This is good news.

At this moment, I want to verify the functionality of the MPL mechanism to resend the information twice. My approach is to run a Python script on OTBR to send multicast packet information and then use a sniffer to check the captured content. However, I couldn't capture any content. On the other hand, when I ping the realm local address, I can capture data. Can you see what might be causing this problem? Perhaps OTBR shouldn't directly send multicast packets? I will be really appreciate if you could give me some advice, thanks!


![1b4672d0cf3a806d5dcbcf80ee98658](https://github.com/openthread/openthread/assets/106933152/b97bf08c-eb67-445f-ab9b-55d9ec2bfdff)

[Uploading multicast.txt…]()

**Reply by jwhui** [reply]
> My approach is to run a Python script on OTBR to send multicast packet information

Can you provide more detail on exactly what you are doing here?

**Reply by ZhouJooJoo** [reply]
hi, what i am trying to do is build a thread network, using OTBR to send multicast packets to Realm local address, to validate that, the MPL do will transfer the packets twice.

**Reply by jwhui** [reply]
Can you provide more detail on exactly what you are doing to send the multicast packets?

**Reply by ZhouJooJoo** [reply]
yes, i would like to see that, the multicast packets will be transfer twice because of the MPL protocol

**Reply by jwhui** [reply]
You said:

> My approach is to run a Python script on OTBR to send multicast packet information and then use a sniffer to check the captured content. However, I couldn't capture any content. 

Unfortunately, without more detail, it is difficult for anyone to debug your issue.

**Reply by ZhouJooJoo** [reply]
this is python script, i am trying to use, use the socket to send a message, and what i found is that, when i try to use the realm local address, as unicast address, the message can be sent successfully, but when i consider it as a multicast address, the packets can not be sent.


```

import socket


unicast_address = 'ff33:0040:fdd9:a608:7b19:9a89:0:1'
port = 10000

try:

    print("Creating UDP socket...")
    sock = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)


    message = "Hello, world!"
    print(f"Sending message: {message} to {unicast_address}:{port}")
    sock.sendto(message.encode(), (unicast_address, port))

    print("Message sent successfully!")

except Exception as e:
    print(f"An error occurred: {e}")

finally:

    print("Closing socket...")
    sock.close()

```

**Reply by jwhui** [reply]
Try configuring the `IPV6_MULTICAST_HOPS` socket option.

**Reply by ZhouJooJoo** [reply]
thanks for the advice! I verified that the python script can do this work, and know i would like to use another way which is more convinient and light to send the message--"using the ot command."
![b2a197750cbb4d041ddd391d9b18990](https://github.com/openthread/openthread/assets/106933152/539d47a7-8378-4e24-b948-1eb0d3475e45)
'1234' is the port number, and i use the wireshark to capture the data, it seems the meessages were sent, however, there are continuous udp messages, which are one byte larger than the normal packet, when i look into the data, it contains one more data, the hoop limit, do you know why is that? or should i use any other OT command to send messages?
![3d54dfded78b41778f2aad05bc3cae2](https://github.com/openthread/openthread/assets/106933152/847f3e76-ac1e-470e-975a-b75f5caaaa9e)
Many thanks!!

**Reply by jwhui** [reply]
Thread forwards IPv6 multicast messages by having the routers transmit them as broadcast messages multiple times. The multicast message should include an IPv6 Hop-by-Hop Option, which allows Thread devices to filter out duplicate receptions.

**Reply by ZhouJooJoo** [reply]
thanks for your reply.  just to clarify that, among the results of packet capture from wireshark above, there are two messages with a blue background, which are exactly what I expected, because the address I sent is the realm local address, and under the action of the MPL mechanism, the information was forwarded twice. But I don’t know why the red packet capture information appears. Based on your suggestions, I modified the sending command as follows, and additionally specified the hop limit. However, there are still red packet capture messages. However, this red packet capture information did not appear when I used a python script to send multicast information. I hope you can provide more ideas, thank you.
the revised command i am using to send the packets is listed below:
`sudo ot-ctl udp send ff33:0040:fdd9:a608:7b19:9a89:0000:0001 1234 hello hlim=50 time=9ms
`

**Reply by jwhui** [reply]
Can you share the pcap file?

**Reply by ZhouJooJoo** [reply]
sure, here is the packet that i tried to capture, and the phenomenon happens at line 1286-1290, 2758-2766
[Uploading nordic TBR.zip…]()

**Reply by jwhui** [reply]
I'm not able to download the packet capture.
